### ISSUE TITLE
Support custom resources in the core install

### ISSUE THREAD

**Issue by olix0r** [linkerd/linkerd2#8173]
Currently, `linkerd install` generates a number of `CustomResourceDefinitions`. This makes it impractical for the command to also generate instances of these resources, as `kubectl apply` does not guarantee that the definitions will be ready before the resource instances are created.

This is a problem, as it makes it difficult for the core control plane to leverage policy resources; and--now that resources `NetworkAuthentication` may be referenced by `AuthorizationPolicies` across namespaces--it would also be convenient for the base install to include default network blocks for, e.g. `kubelet` and `kube-api` that can be managed in one place to configure policies for the entire cluster.

To fix this, I propose that we make the following changes:

* `linkerd install clusterresources` is added, replacing `linkerd install config`.
* `linkerd install` is updated to check that the CRDs are available before generating resources. If they are not, it fails and directs the user to the `linkerd install crds` subcommand.
* `linkerd upgrade` is updated similarly:
  * `linkerd upgrade clusterresources` (replacing `linkerd upgrade config`) always generates the CLI version's CRDs.
  * `linkerd upgrade` fails if the required resource versions are not present in the cluster.
* A check is added via `linkerd check --crds` that ensures the CRDs are installed with at least the CLI's api versions. That is, if the CLI knows about v1beta1 of a resource and the cluster only has `v1alpha1`, the check should fail.
  * `linkerd check --crds --wait=...` should wait for the CRDs to be properly installed.
  * `linkerd check` includes the above CRD check.

This will allow us to update the base install to include custom resources (in followups).

**Comment by alpeb** [comment]
It might simplify things further both code wise and from the user's perspective if we align the CLI with the charts setup. So having `linkerd install crds` replicate what the `linkerd-crds` chart does, and `linkerd install` (with the pre-check you mention) replicate the `linkerd-control-plane` chart. Note that charts don't segregate cluster-level vs namespace-level (that was decided I believe because nobody had really asked for that).

Also note that `linkerd check` has also a `config` subcommand that could be replaced with `linkerd check --crds`.

**Comment by kleimkuhler** [comment]
I'm going to summarize some off-GitHub discussion here. There are two installation paths that we consider are the CLI and Helm; this change would primarily affect the CLI.

The summary is
- CLI installation is split into `linkerd install crds` and `linkerd install control-plane`
- For clusters where a separation of permissions is required, users will be required to use Helm

---

Installation through the CLI is single-step by default through `linkerd install`. It can be a multi-step install through `linkerd install config` and `linkerd install control-plane`. This allows for a separation of cluster permissions, where the resources output by `config` apply cluster-wide and then the resources output by `control-plane` are specific to Linkerd's deployments.

Currently, `ServiceProfiles`, `Servers`, and `ServerAuthorizations` are included in the `config` resources. It will be important that installation of CRDs are isolated from the rest of Linkerd resources so that in the future we have the ability to install cluster-wide CRDs (such as a cluster-wide policy resource).

We then have to distinguish the _other_ resources currently installed by `config` and those installed by `control-plane`. Assuming that we rely off the Helm installation path for clusters where a separation of permissions is important, `control-plane` would now include all non-CRD resources from `config`. This would result in a default two-step CLI installation: `linkerd install crds` and `linkerd install control-plane`.

After the above changes are made, `linked upgrade` would be updated in a similar fashion (as mentioned in the original description) and `linkerd check --crds` would be added.

The ordering of these changes would be
1. Move non-CRD resources in `config` to `control-plane`
2. Deprecate `config` (so that it errors) and introduce `crds` which installs `ServiceProfiles` and all policy-related CRDs
3. Add CRDs check to `control-plane` which will cause it to error if the CRDs are not installed
4. Add similar changes to `linkerd upgrade`
5. Add `linkerd check --crds` that ensures the required CRDs are installed
6. Deprecate `linkerd install` which will result in a _required_ two-step installation

**Comment by adleong** [comment]
My takeaway for our discussions was slightly different.  Namely that we wanted the following:

* A chart for all Linkerd CRDs (currently we only have CRDs for namespaced custom resources, but this may include cluster scoped custom resources in the future)
* A chart for Linkerd cluster scoped resources (excluding CRDs) such as ClusterRoles, etc.
* A chart for namespaced Linkerd resources i.e. the control plane (including Linkerd custom resources instances such as default policy objects)

Mirroring these charts, we would have 3 install subcommands:

* `linkerd install crds` or perhaps `linkerd install --crds` to install the CRDs chart
* `linkerd install clusterresources` (formerly known as install config)
* `linkerd install control-plane` (as it exists today)

And that `linkerd install` combines the clusterresources and control-plane into a single step.  This command should fail with a helpful message if the CRDs have not already been installed.

This results in a 2-step CLI install in the common case (`linkerd install crds` -> `linkerd install`) with the option of doing a 3 step install instead if separation of cluster permissions is desired.  Helm users would always need to install the 3 charts individually, however.

**Comment by alpeb** [comment]
I believe @adleong's summary correctly reflects the current state of the conversation.

You can refer back to #6620 for more context about the original helm charts split (To reiterate, we mentioned there there's another way to include CRDs under the `/templates/crds` for them to be installed before everything else, which would allow us to only have to deal with two charts instead of three, but that sets in stone the CRDs manifests an they can't be upgraded/deleted in future chart releases so we'd be forced to provide instructions for doing so manually. That's why that method was discarded.)

One way we can lower the user's pain of having to deal with three charts is providing default configs and instructions for CD tools. [helmfile](https://github.com/roboll/helmfile) for example is powerful and simple helm wrapper for which we could provide a `helmfile.yaml` config which would group our charts with runtime dependencies (first install crds, then clusterresources, then control-plane).

**Comment by adleong** [comment]
After poking around in the charts, it turns out that splitting the charts up in the way I describe above is rather difficult.  This is because we use Helm certification functions to generate webhook credentials for all of our webhooks and then populate those credentials into the Webhook configuration (cluster scope) and the webhook controllers themselves (namespaced).  This means that those cluster scoped and namespaced resources cannot easily live in different charts.

This makes it difficult to solve the privilege differentiation portion of this in Helm.  However, the primary goal here is to support custom resources in the base install, not privilege differentiation.  To that end and to keep things simple and maintainable, I suggest we actually make the CLI behave more like Helm and split the CRDs into a separate step but put everything else (cluster and namespace scoped resources) all together.

In other words, the `linkerd install --crds` command would install the CRDs only, the `linkerd install` command would install everything else, and the `linkerd install config` and `linkerd install control-plane` subcommands would be removed.

While this removes the ability for the CLI installs to be broken up by privilege level, I think this is the right move since it keeps the install flow simple, easy to understand, and easy to maintain.  Plus it make the CLI behavior match Helm's behavior more closely.
